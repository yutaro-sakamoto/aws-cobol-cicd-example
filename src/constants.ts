/**
 * The name of the ECR repository
 */
export const ecrRepositoryName = 'my-ecr-repo';
/**
 * The SSM parameter name indicating the task definition
 */
export const taskDefinitionArnSsmParamName =
  '/ecs-cicd-example/task-definition-arn';
