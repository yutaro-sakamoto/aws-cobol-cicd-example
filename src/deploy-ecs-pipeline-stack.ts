import * as cdk from "aws-cdk-lib";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as ecs from "aws-cdk-lib/aws-ecs";
//import * as ssm from "aws-cdk-lib/aws-ssm";
import { Construct } from "constructs";

/**
 * DeployEcsPipelineStackのProps
 */
export interface DeployEcsPipelineStackProps extends cdk.StackProps {
  /**
   * ECRリポジトリ名
   */
  ecrRepositoryName: string;
  /**
   *
   */
  fargateService: ecs.FargateService;
}

/**
 * ECSデプロイ用のパイプラインのスタック
 */
export class DeployEcsPipelineStack extends cdk.Stack {
  constructor(
    scope: Construct,
    id: string,
    props: DeployEcsPipelineStackProps,
  ) {
    super(scope, id, props);

    const artifact = new codepipeline.Artifact();

    const service = props.fargateService;

    const pipeline = new codepipeline.Pipeline(this, "Pipeline", {
      pipelineName: "EcsBlueGreenPipeline",
      pipelineType: codepipeline.PipelineType.V2,
    });

    pipeline.addStage({
      stageName: "Source",
      actions: [
        new codepipeline_actions.EcrSourceAction({
          actionName: "ECR",
          repository: ecr.Repository.fromRepositoryName(
            this,
            "ExistingECRRepository",
            props.ecrRepositoryName,
          ),
          imageTag: "latest",
          output: artifact,
        }),
      ],
    });
    pipeline.addStage({
      stageName: "Deploy",
      actions: [
        new codepipeline_actions.EcsDeployAction({
          actionName: "DeployToECS",
          service,
          input: artifact,
        }),
      ],
    });
  }
}
