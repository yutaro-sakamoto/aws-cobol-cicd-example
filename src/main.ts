import { App, Stack, StackProps } from "aws-cdk-lib";
import * as cdk from "aws-cdk-lib";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as ecr from "aws-cdk-lib/aws-ecr";
import { AwsSolutionsChecks, NagSuppressions } from "cdk-nag";
import { Construct } from "constructs";
import * as dotenv from "dotenv";
import * as codebuild from "aws-cdk-lib/aws-codebuild";

dotenv.config();

/**
 * サンプルのスタック
 */
export class EcrStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    new ecr.Repository(this, "ECRRepository", {
      repositoryName: "my-ecr-repo",
    });

    const pipeline = new codepipeline.Pipeline(this, "ApplicationPipeline", {
      pipelineName: "ApplicationPipeline",
    });

    NagSuppressions.addResourceSuppressions(
      pipeline.artifactBucket.encryptionKey!,
      [
        {
          id: "AwsSolutions-KMS5",
          reason: "This key is used for the artifact bucket.",
        },
      ],
    );

    const sourceOutput = new codepipeline.Artifact();

    pipeline.addStage({
      stageName: "Source",
      actions: [
        new codepipeline_actions.CodeStarConnectionsSourceAction({
          actionName: "GitHubSource",
          owner: "yutaro-sakamoto",
          repo: "aws-cobol-cicd-example",
          branch: "dev",
          output: sourceOutput,
          connectionArn:
            "arn:aws:codeconnections:ap-northeast-1:377426933046:connection/7d76c990-5676-472f-9132-70f681587d55",
          triggerOnPush: true,
        }),
      ],
    });

    const buildOutput = new codepipeline.Artifact();
    pipeline.addStage({
      stageName: "Build",
      actions: [
        new codepipeline_actions.CodeBuildAction({
          actionName: "DockerBuild",
          project: new codebuild.PipelineProject(this, "DockerBuildProject", {
            environment: {
              buildImage: codebuild.LinuxBuildImage.AMAZON_LINUX_2_5,
              privileged: true,
            },
            buildSpec: codebuild.BuildSpec.fromObject({
              version: "0.2",
              phases: {
                build: {
                  commands: ["echo hello > message.txt"],
                },
              },
              artifacts: {
                files: ["message.txt"],
              },
            }),
          }),
          input: sourceOutput,
          outputs: [buildOutput],
        }),
      ],
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const stack = new EcrStack(app, "aws-cobol-cicd-example-dev", { env: devEnv });
cdk.Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
NagSuppressions.addStackSuppressions(stack, [
  { id: "AwsSolutions-IAM5", reason: "Allow IAM policies to contain *" },
  {
    id: "AwsSolutions-S1",
    reason: "Server access logs of S3 bucket are unnecessary",
  },
]);
app.synth();
