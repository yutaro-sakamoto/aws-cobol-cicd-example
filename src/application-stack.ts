import { Stack, CfnOutput, StackProps } from "aws-cdk-lib";
import * as codebuild from "aws-cdk-lib/aws-codebuild";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as ecr from "aws-cdk-lib/aws-ecr";
import * as ecs from "aws-cdk-lib/aws-ecs";
import * as iam from "aws-cdk-lib/aws-iam";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { NagSuppressions } from "cdk-nag";
import { Construct } from "constructs";

import * as constants from "./constants";

/**
 * Stack props
 */
export interface ApplicationStackProps extends StackProps {
  /**
   * if true, only synthesize the stack
   */
  synthOnly: boolean;
}

/**
 * サンプルのスタック
 */
export class ApplicationStack extends Stack {
  public readonly ecrRepository: ecr.Repository;
  /**
   * A constructor
   * @param scope scope
   * @param id id
   * @param props properties of this stack
   */
  constructor(scope: Construct, id: string, props: ApplicationStackProps) {
    super(scope, id, props);

    this.ecrRepository = new ecr.Repository(this, "ECRRepository", {
      repositoryName: constants.ecrRepositoryName,
      imageScanOnPush: true,
    });

    const pipeline = new codepipeline.Pipeline(this, "ApplicationPipeline", {
      pipelineName: "ApplicationPipeline",
      pipelineType: codepipeline.PipelineType.V2,
    });

    new CfnOutput(this, "applicationPipelineName", {
      value: pipeline.pipelineName,
    });

    NagSuppressions.addResourceSuppressions(
      pipeline.artifactBucket.encryptionKey!,
      [
        {
          id: "AwsSolutions-KMS5",
          reason: "This key is used for the artifact bucket.",
        },
      ],
    );

    const sourceOutput = new codepipeline.Artifact();

    pipeline.addStage({
      stageName: "Source",
      actions: [
        new codepipeline_actions.CodeStarConnectionsSourceAction({
          actionName: "GitHubSource",
          owner: "yutaro-sakamoto",
          repo: "aws-cobol-cicd-example",
          branch: "dev",
          output: sourceOutput,
          connectionArn:
            process.env.AWS_CODECONNECTIONS_ARN || "aws_codeconections_arn",
          triggerOnPush: false,
        }),
      ],
    });

    const buildOutput = new codepipeline.Artifact();

    pipeline.addStage({
      stageName: "Build",
      actions: [
        new codepipeline_actions.CodeBuildAction({
          actionName: "DockerBuild",
          project: new codebuild.PipelineProject(this, "DockerBuildProject", {
            cache: codebuild.Cache.local(codebuild.LocalCacheMode.DOCKER_LAYER),
            environment: {
              buildImage: codebuild.LinuxBuildImage.AMAZON_LINUX_2_5,
              privileged: true,
              environmentVariables: {
                AWS_ACCOUNT_ID: {
                  value: props.env!.account!,
                },
                AWS_DEFAULT_REGION: {
                  value: props.env!.region!,
                },
                IMAGE_REPO_NAME: {
                  value: constants.ecrRepositoryName,
                },
                IMAGE_TAG: {
                  value: "latest",
                },
              },
            },
            buildSpec: codebuild.BuildSpec.fromObject({
              version: "0.2",
              phases: {
                pre_build: {
                  commands: [
                    "aws ecr get-login-password --region $AWS_DEFAULT_REGION | docker login --username AWS --password-stdin $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com",
                  ],
                },
                build: {
                  commands: [
                    "docker build -t $IMAGE_REPO_NAME:$IMAGE_TAG app",
                    "docker tag $IMAGE_REPO_NAME:$IMAGE_TAG $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG",
                  ],
                },
                post_build: {
                  commands: [
                    "docker push $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG",
                  ],
                },
              },
            }),
            role: new iam.Role(this, "CodeBuildRole", {
              assumedBy: new iam.ServicePrincipal("codebuild.amazonaws.com"),
              managedPolicies: [
                iam.ManagedPolicy.fromAwsManagedPolicyName(
                  "AmazonS3ReadOnlyAccess",
                ),
                iam.ManagedPolicy.fromAwsManagedPolicyName(
                  "AmazonEC2ContainerRegistryPowerUser",
                ),
                iam.ManagedPolicy.fromAwsManagedPolicyName(
                  "CloudWatchLogsFullAccess",
                ),
              ],
            }),
          }),
          input: sourceOutput,
          outputs: [buildOutput],
        }),
      ],
    });
  }
}
