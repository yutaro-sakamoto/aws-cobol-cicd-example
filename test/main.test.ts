import { App, Aspects } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import { EcrStack, InfrastructureStack } from '../src/main';

const app = new App();
const stack = new EcrStack(app, 'ApplicationStack', {
  env: {
    account: 'example-account',
    region: 'example-region',
  },
  synthOnly: true,
});
const infrastructureStack = new InfrastructureStack(app, 'infrastructureStack', {
  env: {
    account: 'example-account',
    region: 'example-region',
  },
  synthOnly: true,
});

Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));

test('Snapshot', () => {

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot('applicationStack');
  const infrastructureTemplate = Template.fromStack(infrastructureStack);
  expect(infrastructureTemplate.toJSON()).toMatchSnapshot('infrastructureStack');
});