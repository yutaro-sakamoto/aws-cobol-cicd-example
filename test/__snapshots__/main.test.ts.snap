// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "applicationPipelineName": {
      "Value": {
        "Ref": "ApplicationPipeline4C988C4B",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApplicationPipeline4C988C4B": {
      "DependsOn": [
        "ApplicationPipelineRoleDefaultPolicyCADAF9C3",
        "ApplicationPipelineRoleADFF393B",
      ],
      "Properties": {
        "ArtifactStore": {
          "EncryptionKey": {
            "Id": {
              "Fn::GetAtt": [
                "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                "Arn",
              ],
            },
            "Type": "KMS",
          },
          "Location": {
            "Ref": "ApplicationPipelineArtifactsBucketD6B45A16",
          },
          "Type": "S3",
        },
        "Name": "ApplicationPipeline",
        "PipelineType": "V2",
        "RoleArn": {
          "Fn::GetAtt": [
            "ApplicationPipelineRoleADFF393B",
            "Arn",
          ],
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": {
                  "BranchName": "dev",
                  "ConnectionArn": "aws_codeconections_arn",
                  "DetectChanges": false,
                  "FullRepositoryId": "yutaro-sakamoto/aws-cobol-cicd-example",
                },
                "Name": "GitHubSource",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_Source_GitHubSource",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleE89A168C",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": {
                  "ProjectName": {
                    "Ref": "DockerBuildProject52E5CDF6",
                  },
                },
                "InputArtifacts": [
                  {
                    "Name": "Artifact_Source_GitHubSource",
                  },
                ],
                "Name": "DockerBuild",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_Build_DockerBuild",
                  },
                ],
                "RoleArn": {
                  "Fn::GetAtt": [
                    "ApplicationPipelineBuildDockerBuildCodePipelineActionRole7DA0A42F",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "ApplicationPipelineArtifactsBucketD6B45A16": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "KMSMasterKeyID": {
                  "Fn::GetAtt": [
                    "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "ApplicationPipelineArtifactsBucketEncryptionKeyAliasF37F2519": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AliasName": "alias/codepipeline-testapplicationpipeline676b7a1a",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
      "UpdateReplacePolicy": "Delete",
    },
    "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834": {
      "DeletionPolicy": "Delete",
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-KMS5",
              "reason": "This key is used for the artifact bucket.",
            },
          ],
        },
      },
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::example-account:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "ApplicationPipelineArtifactsBucketPolicy0F22EB48": {
      "Properties": {
        "Bucket": {
          "Ref": "ApplicationPipelineArtifactsBucketD6B45A16",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ApplicationPipelineArtifactsBucketD6B45A16",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ApplicationPipelineArtifactsBucketD6B45A16",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ApplicationPipelineBuildDockerBuildCodePipelineActionRole7DA0A42F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::example-account:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ApplicationPipelineBuildDockerBuildCodePipelineActionRoleDefaultPolicyA88B4868": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "DockerBuildProject52E5CDF6",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApplicationPipelineBuildDockerBuildCodePipelineActionRoleDefaultPolicyA88B4868",
        "Roles": [
          {
            "Ref": "ApplicationPipelineBuildDockerBuildCodePipelineActionRole7DA0A42F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApplicationPipelineRoleADFF393B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ApplicationPipelineRoleDefaultPolicyCADAF9C3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ApplicationPipelineArtifactsBucketD6B45A16",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ApplicationPipelineArtifactsBucketD6B45A16",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleE89A168C",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineBuildDockerBuildCodePipelineActionRole7DA0A42F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApplicationPipelineRoleDefaultPolicyCADAF9C3",
        "Roles": [
          {
            "Ref": "ApplicationPipelineRoleADFF393B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleDefaultPolicy6ECA3CD7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codestar-connections:UseConnection",
              "Effect": "Allow",
              "Resource": "aws_codeconections_arn",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ApplicationPipelineArtifactsBucketD6B45A16",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ApplicationPipelineArtifactsBucketD6B45A16",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:PutObjectAcl",
                "s3:PutObjectVersionAcl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "ApplicationPipelineArtifactsBucketD6B45A16",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleDefaultPolicy6ECA3CD7",
        "Roles": [
          {
            "Ref": "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleE89A168C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApplicationPipelineSourceGitHubSourceCodePipelineActionRoleE89A168C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::example-account:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "CodeBuildRole728CBADE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonS3ReadOnlyAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonEC2ContainerRegistryPowerUser",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLogsFullAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CodeBuildRoleDefaultPolicy829527DE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:example-region:example-account:log-group:/aws/codebuild/",
                      {
                        "Ref": "DockerBuildProject52E5CDF6",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:example-region:example-account:log-group:/aws/codebuild/",
                      {
                        "Ref": "DockerBuildProject52E5CDF6",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:example-region:example-account:report-group/",
                    {
                      "Ref": "DockerBuildProject52E5CDF6",
                    },
                    "-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "ApplicationPipelineArtifactsBucketD6B45A16",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "ApplicationPipelineArtifactsBucketD6B45A16",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodeBuildRoleDefaultPolicy829527DE",
        "Roles": [
          {
            "Ref": "CodeBuildRole728CBADE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DockerBuildProject52E5CDF6": {
      "Properties": {
        "Artifacts": {
          "Type": "CODEPIPELINE",
        },
        "Cache": {
          "Modes": [
            "LOCAL_DOCKER_LAYER_CACHE",
          ],
          "Type": "LOCAL",
        },
        "EncryptionKey": {
          "Fn::GetAtt": [
            "ApplicationPipelineArtifactsBucketEncryptionKeyD6D77834",
            "Arn",
          ],
        },
        "Environment": {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "EnvironmentVariables": [
            {
              "Name": "AWS_ACCOUNT_ID",
              "Type": "PLAINTEXT",
              "Value": "example-account",
            },
            {
              "Name": "AWS_DEFAULT_REGION",
              "Type": "PLAINTEXT",
              "Value": "example-region",
            },
            {
              "Name": "IMAGE_REPO_NAME",
              "Type": "PLAINTEXT",
              "Value": "my-ecr-repo",
            },
            {
              "Name": "IMAGE_TAG",
              "Type": "PLAINTEXT",
              "Value": "latest",
            },
          ],
          "Image": "aws/codebuild/amazonlinux2-x86_64-standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": true,
          "Type": "LINUX_CONTAINER",
        },
        "ServiceRole": {
          "Fn::GetAtt": [
            "CodeBuildRole728CBADE",
            "Arn",
          ],
        },
        "Source": {
          "BuildSpec": "{
  "version": "0.2",
  "phases": {
    "pre_build": {
      "commands": [
        "aws ecr get-login-password --region $AWS_DEFAULT_REGION | docker login --username AWS --password-stdin $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com"
      ]
    },
    "build": {
      "commands": [
        "docker build -t $IMAGE_REPO_NAME:$IMAGE_TAG app",
        "docker tag $IMAGE_REPO_NAME:$IMAGE_TAG $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG"
      ]
    },
    "post_build": {
      "commands": [
        "docker push $AWS_ACCOUNT_ID.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/$IMAGE_REPO_NAME:$IMAGE_TAG"
      ]
    }
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "ECRRepositoryF3A3832E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "RepositoryName": "my-ecr-repo",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
